amd_openvx_extensions/amd_rpp/source/tensor/ColorCast.cpp

```diff
@@ -0,0 +1,287 @@
+    RpptDescPtr pDstDesc;
+    RpptROI *pSrcRoi;
+    RpptRoiType roiType;
+    vxTensorLayout inputLayout;
+    vxTensorLayout outputLayout;
+    size_t inputTensorDims[RPP_MAX_TENSOR_DIMS];
+    size_t outputTensorDims[RPP_MAX_TENSOR_DIMS];
+};
+
+static vx_status VX_CALLBACK refreshColorCast(vx_node node, const vx_reference *parameters, vx_uint32 num, ColorCastLocalData *data) {
+    vx_status status = VX_SUCCESS;
+
+    // Copy alpha (size N) into first N elements
+    STATUS_ERROR_CHECK(vxCopyArrayRange((vx_array)parameters[4], 0, data->inputTensorDims[0], sizeof(vx_float32), data->pAlpha, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    // Copy rgb floats (size N*3) into first N*3 elements
+    STATUS_ERROR_CHECK(vxCopyArrayRange((vx_array)parameters[3], 0, data->inputTensorDims[0] * 3, sizeof(vx_float32), data->pRgbFloats, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+
+    void *roi_tensor_ptr;
+    if (data->deviceType == AGO_TARGET_AFFINITY_GPU) {
+#if ENABLE_HIP
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[1], VX_TENSOR_BUFFER_HIP, &roi_tensor_ptr, sizeof(roi_tensor_ptr)));
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_BUFFER_HIP, &data->pSrc, sizeof(data->pSrc)));
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[2], VX_TENSOR_BUFFER_HIP, &data->pDst, sizeof(data->pDst)));
+#endif
+    } else if (data->deviceType == AGO_TARGET_AFFINITY_CPU) {
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[1], VX_TENSOR_BUFFER_HOST, &roi_tensor_ptr, sizeof(roi_tensor_ptr)));
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_BUFFER_HOST, &data->pSrc, sizeof(data->pSrc)));
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[2], VX_TENSOR_BUFFER_HOST, &data->pDst, sizeof(data->pDst)));
+    }
+    data->pSrcRoi = reinterpret_cast<RpptROI *>(roi_tensor_ptr);
+
+    // Expand NF layouts: replicate per-sample params and ROI across frames
+    if (data->inputLayout == vxTensorLayout::VX_NFHWC || data->inputLayout == vxTensorLayout::VX_NFCHW) {
+        const unsigned N = static_cast<unsigned>(data->inputTensorDims[0]);
+        const unsigned F = static_cast<unsigned>(data->inputTensorDims[1]); // Num of frames per sample
+        // Expand alpha
+        for (int n = N - 1; n >= 0; n--) {
+            unsigned base = n * F;
+            unsigned srcOff = n * 3;
+            for (unsigned f = 0; f < F; f++) {
+                data->pAlpha[base + f] = data->pAlpha[n];
+                data->pSrcRoi[base + f].xywhROI = data->pSrcRoi[n].xywhROI;
+                unsigned dstIdx = (base + f) * 3;
+                data->pRgbFloats[dstIdx + 0] = data->pRgbFloats[srcOff + 0];
+                data->pRgbFloats[dstIdx + 1] = data->pRgbFloats[srcOff + 1];
+                data->pRgbFloats[dstIdx + 2] = data->pRgbFloats[srcOff + 2];
+            }
+        }
+    }
+
+    // Pack floats into RpptRGB
+    const size_t n = data->pSrcDesc->n;
```

Comment:
we shouldn't do too many data copies in vx_rpp. It should just be a pass through. This code seems un-necessary

Reply:
Done, Made changes to pass vx_tensor for RGB array with pinned memory, avoiding extra copies